File: Linked Lists/Medium/Problem 1/problem1.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

void append(Node **head, int data) {
    Node *newNode = (Node *)malloc(sizeof(Node));

    if (newNode == NULL) {
        printf("ERROR | MALLOC\n");
        return;
    }

    newNode->data = data;
    newNode->next = NULL;

    if (*head == NULL) {
        *head = newNode;
        return;
    }

    for (Node *temp = *head; temp != NULL; temp = temp->next) {
        if (temp->next == NULL) {
            temp->next = newNode;
            return;
        }
    }
}

void betterAppend(Node **head, int *array, int length) {
    for (int i = 0; i < length; i++) {
        append(head, array[i]);
    }
}

void printList(Node *head) {
    for (Node *temp = head; temp != NULL; temp = temp->next) {
        printf("%d->", temp->data);
    }

    printf("NULL\n");
}

void freeList(Node **head) {
    Node *temp = NULL;

    while (*head != NULL) {
        temp = *head;
        *head = (*head)->next;

        free(temp);
    }

    *head = NULL;
}

void swapPair(Node **head) {
    if (*head == NULL) {
        printf("EMPTY LIST\n");
        return;
    }

    if ((*head)->next == NULL) {
        printf("ONLY ONE NODE\n");
        return;
    }

    Node *swapOne = *head;
    Node *swapTwo = (*head)->next;
    Node *restOfList = swapTwo->next;
    Node *previousSwapOneCopy = NULL;

    swapTwo->next = swapOne;
    swapOne->next = restOfList;
    (*head) = swapTwo;
    previousSwapOneCopy = swapOne;

    for (Node *temp = restOfList; temp != NULL; ) {
        swapOne = temp;
        swapTwo = temp->next;
        
        if (restOfList->next == NULL) break;
        else restOfList = swapTwo->next;

        previousSwapOneCopy->next = swapTwo;
        swapTwo->next = swapOne;
        swapOne->next = restOfList;

        previousSwapOneCopy = swapOne;
        temp = restOfList;
    }
}

int main(int argc, char const *argv[]) {
    int array[] = {1, 2, 3, 4, 5, 6};
    int arrayTwo[] = {1, 2, 3, 4, 5};

    Node *head = NULL;

    betterAppend(&head, array, (sizeof(array) / sizeof(array[0])));
    printList(head);
    
    swapPair(&head);
    printList(head);
    
    freeList(&head);

    betterAppend(&head, arrayTwo, (sizeof(arrayTwo) / sizeof(arrayTwo[0])));
    swapPair(&head);
    printList(head);

    freeList(&head);

    return 0;
}

File: Linked Lists/Medium/Problem 1/problem1.py
class Node:
    def __init__(self, data: int):
        self.data: int = data
        self.next: 'Node' = None  # Use forward reference for type hint


class LinkedList:
    def __init__(self):
        self.head: 'Node' = None

    def append(self, data: int) -> None:
        newNode: Node = Node(data)

        if self.head is None:
            self.head = newNode
            return

        temp: Node = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = newNode

    def betterAppend(self, nums: list[int]) -> None:
        for i in nums:
            self.append(i)

    def printList(self) -> None:
        temp: Node = self.head
        while temp is not None:
            print(f"{temp.data}->", end="")
            temp = temp.next
        print("NULL")

    def solution(self) -> None:
        if self.head is None:
            print("EMPTY LIST")
            return

        if self.head.next is None:
            print("ONLY ONE NODE")
            return

        # Swap the first pair and update head
        prev: Node = None
        current: Node = self.head
        self.head = current.next  # New head will be second node

        while current and current.next:
            first: Node = current
            second: Node = current.next
            following: Node = second.next

            # Swap
            second.next = first
            first.next = following

            # Link previous pair's tail to current swapped head
            if prev:
                prev.next = second

            # Move pointers forward
            prev = first
            current = following

    def free(self) -> None:
        self.head = None


def main() -> int:
    listOne: LinkedList = LinkedList()

    # Case 1: Even number of nodes
    listOne.betterAppend([1, 2, 3, 4, 5, 6])
    listOne.printList()
    listOne.solution()
    listOne.printList()
    listOne.free()

    # Case 2: Odd number of nodes
    listOne.betterAppend([1, 2, 3, 4, 5])
    listOne.solution()
    listOne.printList()
    listOne.free()

    # Case 3: Single node
    listOne.betterAppend([42])
    listOne.solution()
    listOne.printList()
    listOne.free()

    # Case 4: Empty list
    listOne.solution()

    return 0


if __name__ == "__main__":
    main()
